### PR TITLE
drivers: modem: Added simcom sim7080 modem driver.

### PR DESCRIPTION
**Pull request by lgehreke**
Implemented driver for the simcom sim7080 modem.
This driver features Socket offloading, TCP, UDP, DNS,
SMS, GPS and FTP.

### CODE REVIEW

**Comment by beriberikix** [comment]
Awesome! Do you have a targeted board that can be used for documentation?

**Comment by lgehreke** [comment]
We currently use the sim7080 on a self designed board with a nrf52833. Maybe the `nrf52833dk_nrf52833` could be used for documentation.

**Comment by parthitce** [comment]
@lgehreke I have a breakout board from waveshare. Does it makes sense to add it as shield?

**Comment by beriberikix** [comment]
Ya, waveshare is the vendor that I see that the most. Here's their [SIM7000G](https://www.waveshare.com/product/iot-communication/long-range-wireless/nb-iot-lora/sim7000g-nb-iot-hat.htm) & [SIM7070g](https://www.waveshare.com/product/iot-communication/long-range-wireless/sim7070g-cat-m-nb-iot-gprs-hat.htm).

**Comment by lgehreke** [comment]
Then it would be most reasonable to document this driver with a waveshare shield.

File: drivers/modem/simcom-sim7080.c [hunk 1]
```diff
@@ -0,0 +1,2376 @@
...
+ * Connects an modem socket. Protocol can either be TCP or UDP.
+ */
+static int offload_connect(void *obj, const struct sockaddr *addr, socklen_t addrlen)
+{
+	struct modem_socket *sock = (struct modem_socket *)obj;
+	uint16_t dst_port = 0;
+	char *protocol;
+	struct modem_cmd cmd[] = { MODEM_CMD("+CAOPEN: ", on_cmd_caopen, 2U, ",") };
+	char buf[sizeof("AT+CAOPEN: #,#,#####,####.###.###.####,####")];
+	char ip_str[NET_IPV4_ADDR_LEN];
```
Comment (rerickson1) [inline drivers/modem/simcom-sim7080.c:131]: Do you only support IPv4? This looks like it may not work for IPv6.

**Review by rerickson1 — CHANGES_REQUESTED** [review]
You should add this driver to tests/drivers/build_all/modem

File: drivers/modem/simcom-sim7080.c (hunk 1 above)
Comment (lgehreke) [reply drivers/modem/simcom-sim7080.c:131]: Yes, currently only ipv4 is supported

Comment (lgehreke) [reply drivers/modem/simcom-sim7080.c]: Expanded the buffer to support ipv6

**Comment by beriberikix** [comment]
We also grabbed a few of these [ESP32 + SIM7000G boards](http://www.lilygo.cn/claprod_view.aspx?TypeId=21&Id=1297&FId=t28:21:28) that we hope to target once this is merged!

**Review by rerickson1 — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

File: drivers/modem/Kconfig.simcom-sim7080 [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	select NET_SOCKETS_OFFLOAD
+	imply GPIO
+	help
+	  Enables the driver for the Sim7080 modem.
+
+if MODEM_SIM7080
+
+config MODEM_SIMCOM_SIM7080_RX_STACK_SIZE
+	int "Stack size for the simcom sim7080 modem driver rx thread"
+	default 1028
```
Comment (ycsin) [inline drivers/modem/Kconfig.simcom-sim7080:22]: wouldn't this look better if it is 1024?

File: drivers/modem/Kconfig.simcom-sim7080 [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	int "simcom sim7080 driver init priority"
+	default 80
+	help
+	  simcom sim7080 druver initialization priority.
+
+config MODEM_SIMCOM_SIM7080_LTE_BANDS
+	string "LTE bands the driver can use"
+	default "8,20,28"
+	help
+		Comma separated list of usable lte bands.
```
Comment (ycsin) [inline drivers/modem/Kconfig.simcom-sim7080:42]: some of the help's indentation look off

File: drivers/modem/simcom-sim7080.h [hunk 4]
```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (C) 2021 metraTec GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef SIMCOM_SIM7080_H
+#define SIMCOM_SIM7080_H
+
+#include <kernel.h>
```
Comment (ycsin) [inline drivers/modem/simcom-sim7080.h:10]: I believe some of these header are not required.

A lot of headers are included when you include `zephyr.h`, i.e. `kernel.h` and probably `errno.h`

File: drivers/modem/simcom-sim7080.c [hunk 5]
```diff
@@ -0,0 +1,2384 @@
...
+	dns_result.ai_family = AF_INET;
+	dns_result_addr.sa_family = AF_INET;
+	dns_result.ai_addr = &dns_result_addr;
+	dns_result.ai_addrlen = sizeof(dns_result_addr);
+	dns_result.ai_canonname = dns_result_canonname;
+	dns_result_canonname[0] = '\0';
+
+	if (service) {
+		port = atoi(service);
+		if (port < 1 || port > USHRT_MAX)
```
Comment (ycsin) [inline drivers/modem/simcom-sim7080.c:694]: I believe this is not quite the convention of Zephyr?

File: drivers/modem/simcom-sim7080.c [hunk 6]
```diff
@@ -0,0 +1,2384 @@
...
+const struct socket_dns_offload offload_dns_ops = {
+	.getaddrinfo = offload_getaddrinfo,
+	.freeaddrinfo = offload_freeaddrinfo,
+};
+
+static struct net_if_api api_funcs = {
+	.init = modem_net_iface_init,
+};
+
+static bool offload_is_supported(int family, int type, int proto)
```
Comment (ycsin) [inline drivers/modem/simcom-sim7080.c:747]: I'm not sure if adding `inline` for short function like this improve the performance

File: drivers/modem/simcom-sim7080.c [hunk 7]
```diff
@@ -0,0 +1,2384 @@
...
+ *
+ * The urc has the form +APP PDP: <index>,<state>.
+ * State can either be ACTIVE for activation or
+ * DEACTIVE if disabled.
+ */
+MODEM_CMD_DEFINE(on_urc_app_pdp)
+{
+	mdata.pdp_active = strcmp(argv[1], "ACTIVE") == 0;
+	LOG_INF("PDP context: %u", mdata.pdp_active);
+	k_sem_give(&mdata.sem_response);
```
Comment (ycsin) [inline drivers/modem/simcom-sim7080.c:811]: I think you might need to set the error code if you want to give `sem_response` here?

**Comment by ycsin** [comment]
Some comments when I was referencing this driver.
Also I wonder if new hardware support should be backported? @cfriedt

**Pull request merged**